hibernate-core/src/main/java/org/hibernate/type/BasicTypeRegistry.java

```diff
@@ -205,14 +209,28 @@ private <J> BasicType<J> createIfUnregistered(
 		if ( registeredTypeMatches( javaType, jdbcType, registeredType ) ) {
 			return registeredType;
 		}
-		else {
-			final BasicType<J> createdType = creator.get();
-			register( javaType, jdbcType, createdType );
-			return createdType;
+		// Create an ad-hoc type since the java type doesn't come from the registry and is probably explicitly defined
+		else if ( typeConfiguration.getJavaTypeRegistry().resolveDescriptor( javaType.getJavaType() ) == javaType ) {
+			final var basicTypeReferences = typeReferencesByJavaTypeName.get( javaType.getTypeName() );
+			if ( basicTypeReferences != null && !basicTypeReferences.isEmpty() ) {
+				final var jdbcTypeRegistry = typeConfiguration.getJdbcTypeRegistry();
+				for ( var typeReference : basicTypeReferences ) {
+					if ( jdbcTypeRegistry.getDescriptor( typeReference.getSqlTypeCode() ) == jdbcType ) {
+						final var basicType = typesByName.get( typeReference.getName() );
+						//noinspection unchecked
+						return registeredTypeMatches( javaType, jdbcType, basicType )
+								? (BasicType<J>) basicType
+								: (BasicType<J>) createBasicType( typeReference.getName(), typeReference );
```

Comment:
I just lost a week of my life getting rid of all the dangerous unchecked typecasts in this and related classes, and in the process found bugs.

If you're going to do an unchecked cast, please check it first. Otherwise you've opened the door for the bugs to come seeping back in.